rclcpp/src/rclcpp/executors/detail/timer_manager.hpp

```diff
@@ -0,0 +1,356 @@
+#pragma once
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <vector>
+#include <map>
+
+#include <rcl/timer.h>
+#include <rclcpp/timer.hpp>
+
+#include <inttypes.h>
+
+namespace rclcpp::executors
+{
+
+/**
+ * @brief A class for managing a queue of timers
+ *
+ * This class holds a queue of timers of one type (RCL_ROS_TIME, RCL_SYSTEM_TIME or RCL_STEADY_TIME).
+ * The queue itself manages an internal map of the timers, orders by the next time a timer will be
+ * ready. Each time a timer is ready, a callback will be called from the internal thread.
+ */
+class TimerQueue
+{
+  struct TimerData
+  {
+    std::shared_ptr<const rcl_timer_t> rcl_ref;
+    std::function<void()> timer_ready_callback;
+  };
+
+public:
+  TimerQueue(rcl_clock_type_t timer_type)
+  : timer_type(timer_type),
+    used_clock_for_timers(timer_type),
+    trigger_thread([this]() {
+        timer_thread();
+      })
+  {
+  };
+
+  ~TimerQueue()
+  {
+    stop();
+
+    trigger_thread.join();
+  }
+
+  void stop()
+  {
+    running = false;
+    used_clock_for_timers.cancel_sleep_or_wait();
+  }
+
+  /**
+   * @brief Removes a new timer from the queue.
+   * This function is thread safe.
+   *
+   * Removes a timer, if it was added to this queue.
+   * Ignores timers that are not part of this queue
+   *
+   * @param timer the timer to remove.
+   */
+
+  void remove_timer(const rclcpp::TimerBase::SharedPtr & timer)
+  {
+    rcl_clock_t * clock_type_of_timer;
+
+    std::shared_ptr<const rcl_timer_t> handle = timer->get_timer_handle();
+
+      if (rcl_timer_clock(
+        const_cast<rcl_timer_t *>(handle.get()),
+        &clock_type_of_timer) != RCL_RET_OK)
+    {
+      assert(false);
+    }
+
+    if (clock_type_of_timer->type != timer_type) {
+      // timer is handled by another queue
+      return;
+    }
+
+    timer->clear_on_reset_callback();
```

Comment:
we should first check that the timer is managed by this queue (as the timer could be from another executor) before clearing the callback